python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py

```diff
@@ -0,0 +1,350 @@
+from typing import Generator, Tuple
+
+import pytest
+from dagster import (
+    AssetCheckResult,
+    AssetCheckSpec,
+    AssetExecutionContext,
+    AssetKey,
+    AssetOut,
+    AssetSpec,
+    IOManager,
+    MaterializeResult,
+    StaticPartitionsDefinition,
+    asset,
+    build_op_context,
+    instance_for_test,
+    materialize,
+    multi_asset,
+)
+from dagster._core.errors import DagsterInvariantViolationError, DagsterStepOutputNotFoundError
+from dagster._core.storage.asset_check_execution_record import AssetCheckExecutionRecordStatus
+
+
+def _exec_asset(asset_def, selection=None, partition_key=None):
+    result = materialize([asset_def], selection=selection, partition_key=partition_key)
+    assert result.success
+    return result.asset_materializations_for_node(asset_def.node_def.name)
+
+
+def test_materialize_result_asset():
+    @asset
+    def ret_untyped(context: AssetExecutionContext):
+        return MaterializeResult(
+            metadata={"one": 1},
+        )
+
+    mats = _exec_asset(ret_untyped)
+    assert len(mats) == 1, mats
+    assert "one" in mats[0].metadata
+    assert mats[0].tags
+
+    # key mismatch
+    @asset
+    def ret_mismatch(context: AssetExecutionContext):
+        return MaterializeResult(
+            asset_key="random",
+            metadata={"one": 1},
+        )
+
+    with pytest.raises(
+        DagsterInvariantViolationError,
+        match="Asset key random not found in AssetsDefinition",
+    ):
+        materialize([ret_mismatch])
+
+    @asset
+    def ret_two():
```

Comment:
should returning two `MaterializeResult`s work? seems like odd behavior to me, but this test passes

Reply:
i'm surprised to see this works.

`MaterializeResult` now reads:
> An object representing **a successful materialization of an asset**. These can be returned from
    @asset and @multi_asset decorated functions to pass metadata or specify specific assets were
    materialized.

given that semantic, i think we should restrict `MaterializeResult` to be 1:1 with asset.